types/dispatcher.d.ts

```diff
@@ -215,13 +215,52 @@ declare namespace Dispatcher {
     context: object;
   }
   export type StreamFactory = (data: StreamFactoryData) => Writable;
+  export interface Controller {
+    readonly aborted: boolean;
+    readonly reason: Error;
+    readonly paused: boolean;
+
+    pause(): void;
+    resume(): void;
+    abort(reason: Error): void;
+  }
   export interface DispatchHandlers {
+    // Existing API
+
+    /** Invoked when request is upgraded either due to a `Upgrade` header or `CONNECT` method. */
+    onUpgrade?(statusCode: number, headers: Buffer[] | string[] | null, socket: Duplex): void;
+
+    // New API
+
+    /** Invoked when request is dispatched on a dispatcher */
+    onDispatch?(controller: Controller): void;
+    /** Invoked before request is starting to be processed */
+    onRequestStart?(): void;
+    /** Invoked before headers data is sent */
+    onRequestHeaders?(headers: Array<Buffer|string>): void;
+    /** Invoked before payload data is sent. */
+    onRequestData?(chunk: Buffer | string): void;
+    /** Invoked after request has finished sending */
+    onRequestEnd?(): void;
+    /** Invoked after request has errored while sending */
+    onRequestError?(err: Error): void;
+    /** Invoked before response is starting to be processed */
+    onResponseStart?(): void;
+    /** Invoked after status headers data has been received */
+    onResponseHeaders?(headers: Array<Buffer|string>, statusCode: number, statusText: string): void;
```

Comment:
Nit: remember that technically the `statusText` is optional. So it should be `statusText?: string`.